src/relay/pass/fold_constant.cc

```diff
@@ -15,21 +15,22 @@ using FInterpreter = runtime::TypedPackedFunc<Value(Expr)>;
 
 class ConstantChecker : private ExprVisitor {
  public:
-  // Check whether an expression is constant. The results are memorized.
+  // Check whether an expression is constant. The results are memoized.
   bool Check(const Expr& expr) {
-    if (expr.as<ConstantNode>()) {
-      return true;
-    }
     const auto it = memo_.find(expr);
     if (it != memo_.end())
       return it->second;
     VisitExpr(expr);
-    return memo_[expr];  // return memorized result or the default value false
+    return memo_[expr];  // return memoized result or the default value false
   }
 
  private:
   std::unordered_map<Expr, bool, NodeHash, NodeEqual> memo_;
 
+  void VisitExpr_(const ConstantNode* n) final {
+    memo_[GetRef<Constant>(n)] = true;
```

Comment:
Is there a test case where the previous default behavior fails? This would be a good regression to add a test for.

Reply:
I added this to the patch, mainly to discuss it.

I believe the code before is semantically equivalent to what I've changed it to, so I'm curious why the `ConstantNode` case was baked into `Check`.  Is it just for performance purposes?  If not, then I think this change just increases readability.

Reply:
Oh, I see. It looks like it might be to avoid having constant nodes in the map (trying to reduce memory usage because constant nodes are a trivial case?). Not sure whether the change is necessary.

Reply:
Yeah, I'm pretty sure it's an optimization now, so I don't think we should change the actual code.  But in the tutorial, I'm thinking about having the code snippet differ from the actual code, just for for pedagogical reasons.